spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/source/SparkScanBuilder.java

```diff
@@ -371,12 +376,45 @@ private Schema schemaWithMetadataColumns() {
             .distinct()
             .map(name -> MetadataColumns.metadataColumn(table, name))
             .collect(Collectors.toList());
-    Schema meta = new Schema(fields);
+    // only calculate potential column id collision if meta column was requested
+    Schema meta = fields.isEmpty() ? new Schema(fields) : calculateUnionedSchema(fields);
 
     // schema or rows returned by readers
     return TypeUtil.join(schema, meta);
   }
 
+  private Schema calculateUnionedSchema(List<Types.NestedField> metaColumnFields) {
+    // Calculate used ids by union metadata columns with all base table schemas
+    Set<Integer> currentlyUsedIds =
```

Comment:
yeah I get your point, I think for currentlyUsedIds we only want to keep the reserved field Ids like 2147483642 for `_partitions` but not for inner nested field id. We can try to use `TypeUtil.selectNot` for removing each of field id in

```java
Partitioning.partitionType(table)
.fields().stream()
.map(Types.NestedField::fieldId)
.collect(Collectors.toList())
```

But I feel like we are going to union with all historical table schemas as well, so even if we conflicted it will be reduced.

Reply:
yeah actually on a second thought I can actually drop those nested field in one line so updated here per your suggestion